core/src/main/scala/kafka/server/KafkaApis.scala

```diff
@@ -1446,11 +1448,20 @@ class KafkaApis(val requestChannel: RequestChannel,
       groupCoordinator.handleHeartbeat(
         heartbeatRequest.data.groupId,
         heartbeatRequest.data.memberId,
+        getGroupInstanceId(heartbeatRequest.data.groupInstanceId),
         heartbeatRequest.data.generationid,
         sendResponseCallback)
     }
   }
 
+  def getGroupInstanceId(rawInstanceId: String): Option[String] = {
+      if (rawInstanceId == null ||
+        config.interBrokerProtocolVersion < KAFKA_2_3_IV0)
```

Comment:
I wonder if it's sufficient to do this for JoinGroup only. Basically we just try to gate entrance into the group.

By the way, it's worth add a comment in the JoinGroup handler explaining the purpose of the IBP check. Took me a little while to recall that it was tied to the schema we use in the offsets topic.

Reply:
This pre-check should keep the behavior in `GroupCoordinator` consistent IMO.

Reply:
The problem is that a static member which already received a JoinGroup response could be downgraded to a dynamic member. Have you thought through what the implications of this are?

Reply:
We have IBP to guard against turning a member into static member. So as long as IBP < 2.3, we will not see a downgrade correct?